### ISSUE TITLE
enter Normal Mode with a sequence

### ISSUE THREAD

**Issue by godbout** [godbout/kindaVim.theapp#49]
1. a sequence of letters to enter Normal Mode
2. typing the sequence shouldn't slow down the normal typing experience (not delay after first key of sequence is entered)

**Comment by mrjones2014** [comment]
Would love to be able to use `jk`

**Comment by godbout** [comment]
nice. thanks for letting me know. i'm on #62 for the next release but i'll (re-)dig into the sequence right after.

i had it working well on text inputs a few months ago. the issue is with UI Elements. there's two ways to implement this:
1. when you press `j`, wait a certain amount of time and cancel the `j` (never appears) if `k` is pressed within the interval. that's the common way it's done around, but i loathe it. it slows down every `j` press, i find it very noticeable even with a short interval. it's really bad UX. advantage though is that the sequence could be configurable.
2. delete the `j` if `k` is pressed with the interval. that was working well with text, but i'm not sure what's the best for UI Elements. just let the `j` go through and enter Normal Mode? let the `j` go through, apply a `k`, and enter Normal Mode?

any thoughts/preferences on how you'd like to see this working?

to give a concrete example. let's say you're in Airmail. `jk` in Airmail go down and up the list of emails. you press `jk` within the interval. should Airmail go down one email and enter Normal Mode? should it go one email down, one email up and enter Normal Mode? (not doing anything at all would require the slow bad UX, which i want to avoid.)

**Comment by mrjones2014** [comment]
I agree that I loathe the first way you've listed. In Neovim I use this plugin to make it work essentially like the second way you've listed: https://github.com/max397574/better-escape.nvim

Maybe you can get some inspiration from there.

I think having `jk` work to go to normal mode in UI elements other than text inputs is less important, like you say I'd want to prefer `j`/`k` working for moving up and down over going to normal mode. Is there anything you can do from normal mode when not in a text input, other than go to command mode?

**Comment by godbout** [comment]
> I agree that I loathe the first way you've listed. In Neovim I use this plugin to make it work essentially like the second way you've listed: https://github.com/max397574/better-escape.nvim
> 
> Maybe you can get some inspiration from there.

nice, thanks. looks like what i had done for within text inputs. [NeoVintageous](https://github.com/NeoVintageous/NeoVintageous) for Sublime Text works the same way. still need to rack my brains for UI Elements.

> I think having `jk` work to go to normal mode in UI elements other than text inputs is less important, like you say I'd want to prefer `j`/`k` working for moving up and down over going to normal mode.

so how would you go into Normal Mode if you're not in text? `jk` within text, another way (escape/custom shortcut) when not in text? or are you planning to use kV only within texts?

the idea of kV—at least currently—is that the normal macOS is your Insert Mode. then you enter Normal Mode through `escape` or your Custom Shortcut, and you're in Normal Mode whether you're in text or on some UI Elements. then you do your move, and kV translates depending on the context. pressing `w` in programmatically readable in text? calculate the beginning of the next word. in a text non readable? remap to `alt right alt right alt left`. on a menu? remap to `left` etc etc. having kV entering Normal Mode depending on the context seems very confusing to me tbh. 

> Is there anything you can do from normal mode when not in a text input, other than go to command mode?

not sure if you mean currently, or something i could do.
currently yes, UI Elements are part of the deal. for example in Mail you can navigate through your emails, enter Visual Mode with `v` or `V`, select emails, delete them with `d` or `x`, or delete and go back to Insert Mode with `c`, etc.
if you mean what could be done, well, i guess basically anything 😅  but as said above, currently kV is built as 1) you enter Normal Mode and do your Vim move 2) kV checks the context and translates the move accordingly. not the opposite like 1) in which context are you 2) then do different Vim stuff. hope that makes sense.

**Comment by mrjones2014** [comment]
Gotcha, yeah I think that all makes sense. I'm not sure how to handle it for UI elements.

**Comment by godbout** [comment]
👍 no worries. i'll see when i start digging again. probably by next week. maybe that option will start as beta and you try and see, give feedback etc. i release quite often so we can play and tweak. will ping once i start.

**Comment by godbout** [comment]
kV16 out: https://github.com/godbout/kindaVim.theapp/releases/tag/16

so in text inputs the `j` will be deleted when you press the `k` fast enough. in UI, the `j` will be caught by macOS first. see how it works out for you. the other user would prefer a delay. so i may add it as an option later. but as usual everything is done through automated tests, which means work and time needed, not just quick tweaks. anyways, this is the first version. i'll be adapted according to you guys' feedback. sequence and delay are configurable.

**Comment by mrjones2014** [comment]
Oh man this is amazing. It works great. I'll keep using it and let you know if I run into any issues.

**Comment by godbout** [comment]
actually right after the release i've just found the first letter of the sequence isn't getting deleted in input text that don't handle the Accessibility 🥺️

(also careful with entering your sequence. there's an easter egg with the sequence field. it's going to use the delay you set and if you do the sequence below that limits kV goes into Insert Mode instead of registering your sequence 😅️)

**Comment by godbout** [comment]
> actually right after the release i've just found the first letter of the sequence isn't getting deleted in input text that don't handle the Accessibility 🥺️

ok i know where it's coming from. fix will come in next release.

**Comment by mrjones2014** [comment]
> if you do the sequence below that limits kV goes into Insert Mode instead of registering your sequence 😅️

FWIW I haven't run into this issue

**Comment by godbout** [comment]
> FWIW I haven't run into this issue

easter egg total failure then ☹️

let me know how it goes during these days. i'm gonna work on the fixes and adding more tests, and hopefully by the next release we can close. this part is touching the Core and because it's dealing with event taps it's pretty painful to test. not the most fun i had on kV 😅️ thanks!

**Comment by mrjones2014** [comment]
Okay I did notice today, in some Electron apps, typing `jk` goes to normal mode but doesn't delete the `j`.

**Comment by godbout** [comment]
can you let me know which app, and whether you put the app in a specific Family? that'd help me knowing if it's related to the issue talked about above (and fixed) or if this is a new issue with Electron and the Accessibility. (could be. already a few hacks to handle delete and paste in normal times. if it's that case that's gonna be a painful one.)

thanks!

**Comment by mrjones2014** [comment]
It seems to happen with all Electron apps, even when added to the Electron family.

**Comment by godbout** [comment]
there's two cases for Electron apps:
1. Electron apps that send back proper Accessibility info. those ones need to be put in the Electron Family
2. Electron apps that send wrong Accessibility info. those ones need to be put in the Key Mapping Family

i've tried a couple of ones (Signal, WhatsApp), they're all sending back wrong info, so i put them into Key Mapping. for those ones, the fix that will be available in next release works. see below:

https://user-images.githubusercontent.com/121373/163333018-91aca49f-95ff-4a87-b551-1de5346ecc81.mp4

but i'm not sure with Electron apps that send back correct Accessibility info. do you have any example? sending back correct Accessibility info would mean that when you use kV, it works well and has the block cursor.

**Comment by godbout** [comment]
also, regarding the current implementation, that is the 1st one below:

> 1. when you press `j`, wait a certain amount of time and cancel the `j` (never appears) if `k` is pressed within the interval. that's the common way it's done around, but i loathe it. it slows down every `j` press, i find it very noticeable even with a short interval. it's really bad UX. advantage though is that the sequence could be configurable.
> 2. delete the `j` if `k` is pressed within the interval. that was working well with text, but i'm not sure what's the best for UI Elements. just let the `j` go through and enter Normal Mode? let the `j` go through, apply a `k`, and enter Normal Mode?

this is going to bring some UX issue when dealing with UI elements. for text inputs, i delete the first key of the sequence. for UI elements, i don't. so using a sequence in UI elements might "cancel" what you're trying to do.

example in Airmail:

1. entering Normal Mode
2. selecting an email
3. bringing the snoozing option with `⌘d`
4. entering Normal Mode
5. choosing a snooze time

the result in my case, using `escape` (caps lock, remapped through macOS default keyboard options) to enter Normal Mode:

https://user-images.githubusercontent.com/121373/163335147-755852eb-52a9-48a9-b2cf-4a3bcb043c9c.mp4

the result using `jk` to enter Normal Mode:

https://user-images.githubusercontent.com/121373/163335215-c5ee4f35-32a8-4cfe-91c0-671cde6d2852.mp4

you can't choose the snoozing time because the first `j` will cancel/close the box.

the only way to handle this would be to go through solution 2 quoted above. having the first character of the sequence fire ONLY after the delay. i guess that could be done only when we deal with UI elements, not with text. so within text your typing would never been slowed down/never feel sticky. but it would with UI elements. (also tbh that would be painful to implemented, and to test—not even sure automated test would be possible here.)

we can't go around this. the usual solutions like the one you referred to before (better-escape) only deal with text and/or their own buffer. kV brings Vim Mode everywhere, text and UI, and doesn't own its own buffer. so we're stuck here.

any thoughts? personally i'm not gonna use the sequence. i enter with `escape`. so i'm ok with the current situation (with the fix).

**Comment by mrjones2014** [comment]
>but i'm not sure with Electron apps that send back correct Accessibility info. do you have any example? sending back correct Accessibility info would mean that when you use kV, it works well and has the block cursor in that app.

I'm not sure. The only Electron apps I use are Slack, Signal, Discord, Notion, and 1Password. Those all seem to fall in the Electron family. However, I noticed that Discord, 1Password, and Notion don't seem to work at all. I kinda actually get why Notion and 1Password wouldn't work (the "block" editing mechanics, and possibly some security measures, respectively). Discord I would expect to work though. In Notion and Discord you can enter normal mode, but then nothing else works. In 1Password I can't even enter normal mode.

It also seems like sometimes entering normal mode in Discord sometimes crashes kindaVim entirely 😅 

I think I'd also be fine with adding the delay for the `j` insertion, since I keep my delay setting super low (I've been tweaking it but it's between 100-140ms).

**Comment by mrjones2014** [comment]
Actually, it looks like dropping Notion and Discord in the "Key Mapping" family under advanced makes them work. 1Password still won't even let me enter normal mode. I'm kinda fine with that though as I'm guessing it's due to some extra security measures that 1Password's desktop app team is taking.

**Comment by godbout** [comment]
> I'm not sure. The only Electron apps I use are Slack, Signal, Discord, Notion, and 1Password. Those all seem to fall in the Electron family. However, I noticed that Discord, 1Password, and Notion don't seem to work at all. I kinda actually get why Notion and 1Password wouldn't work (the "block" editing mechanics, and possibly some security measures, respectively). Discord I would expect to work though. In Notion and Discord you can enter normal mode, but then nothing else works. In 1Password I can't even enter normal mode.

ok, i'll check all of them. 
what i can tell you for now based on my past tests:
1. Slack sends back wrong Accessibility data. for example it's mixing caret location with line number. so if in a text your caret location is at character 10, that is on line 2, Slack will tell you you're on character 2. so, impossible to know where you are properly, and impossible to move accurately. so Slack needs to go to Key Mapping, where Vim Mode is handled through key remapping, not text manipulation as text manipulation is not possible due to the wrong data. so you'll be able to do some stuff, like move by word, go to the end of file line, screen line, beginning or end of text etc. you'll be able to use the count system for some moves too. but everything that requires text calculation will not work. current Key Mapping (Keyboard Strategy) implementation is here: https://github.com/godbout/kindaVim.theapp#kv-keyboard-strategy-keyboard-remapping-less-accurate2. 
2. Signal sends back wrong data too. needs to be put also in Key Mapping
3. 1Password i guess you mean 1Password 8? 1Password 7 works even if the fields are Secure Input Fields. when i tried 1Password 8 i wasn't even able to get the Viming Window working correctly. seems the Accessibility API cannot do anything with 1Password 8. unfortunately that's the switch to Electron issue. will investigate further, but that's where the limits of macOS APIs and Electron APIs are gonna be reached

i'll check Notion and Discord.
for all the Electron apps that don't work when you move them into the Electron Family, you need to move them to the Key Mapping Family.
as said earlier, unfortunately, the Electron ecosystem is a huge mess. different apps use different versions of the framework. then different apps use or custom the Accessibility the way they want. nothing is standard. only way to make those app work is to ask the developers to make them accessible properly.

> It also seems like sometimes entering normal mode in Discord sometimes crashes kindaVim entirely 😅

shit. probably even worse than expected then, and sending really, really wrong data. will investigate.

> I think I'd also be fine with adding the delay for the `j` insertion, since I keep my delay setting super low (I've been tweaking it but it's between 100-140ms).

do you mean for the UI elements only, or also within input text? (that'll help if we use the same system for both.)

**Comment by godbout** [comment]
> Actually, it looks like dropping Notion and Discord in the "Key Mapping" family under advanced makes them work. 1Password still won't even let me enter normal mode. I'm kinda fine with that though as I'm guessing it's due to some extra security measures that 1Password's desktop app team is taking.

with 1Password 8 i was able to enter Normal Mode and go up/down the list of items, but can't do any moves yeah. first app that i can't get the Viming window to work also. float panels are handled, like Alfred, RayCast etc. i'm afraid that unfortunately 1Password 8's new core and UI framework that works on every platform works on the surface, but not when you need specific, deep APIs :/

**Comment by mrjones2014** [comment]
Hm. Signal and Slack both seem to work fine for me when just added to the Electron family (as opposed to key mapping family). I'll try both.

>1Password i guess you mean 1Password 8?

Yeah my bad, I meant 1Password 8.

>unfortunately, the Electron ecosystem is a huge mess

Yeah that's definitely true 😅 

>do you mean for the UI elements only, or also within input text? (that'll help if we use the same system for both.)

I think it'll be fine for both. My delay setting is so low anyway, I don't know if I'd even notice the delay while typing. Also the delay would sorta be cancelled if I type `j` and then another letter that isn't `k` right? Like if I'm typing a normal word like `jobs`, there wouldn't even really be a delay once I press the `o` key? Regardless I think the delay approach is fine by me.

**Comment by godbout** [comment]
> Hm. Signal and Slack both seem to work fine for me when just added to the Electron family (as opposed to key mapping family). I'll try both.

how much do you want to know? i can explain more (i want to) but not sure you're interested and/or want to read/understand what's happening behind the scenes. (in Signal—latest version anyway—

> Yeah my bad, I meant 1Password 8.

ok. i'll dig again. 

> Yeah that's definitely true 😅

yeah. that's what i wanted to mean earlier when i said that this is where we're gonna reach the bumpy road or something like that. kV works very well and is consistent, until unfortunately you start playing with bad actors. and the Electron apps are the bad actors (personally i use none). some native apps implement the AX, some native apps don't. those are very clear and how to handle them is controllable. Electron apps are not.

> I think it'll be fine for both. My delay setting is so low anyway, I don't know if I'd even notice the delay while typing. Also the delay would sorta be cancelled if I type `j` and then another letter that isn't `k` right? Like if I'm typing a normal word like `jobs`, there wouldn't even really be a delay once I press the `o` key? Regardless I think the delay approach is fine by me.

correct, the delay would be cancelled. but you would press `j` and then `o` and `jo` would kinda appear together. `j` would appear RIGHT after you pressed `j`. might be fine for others. just not for me 😀️

**Comment by godbout** [comment]
> ok. i'll dig again.

ok, regarding 1Password 8. nothing personal, but i think this is exactly why people complained. the Electron stuff is just a fucking nightmare for Accessibility.

you'll have to move 1Password 8 to the `911 Family`. the reason is:

https://user-images.githubusercontent.com/121373/163409630-6326d82b-06b1-493b-a57b-784a8eb39110.mp4

the whole thing is just a big ScrollView. the "workflow" with Electron apps is 1) put it in the `Electron Family` first. it will activate the Accessibility of the Electron framework and then it'll work if the Electron app sends back correct data. if it doesn't (motions don't work properly), then move the app to the `Key Mapping Family`. in there kV will use key remapping. it will detect whether you're in a text input or not, because the key remapping will be different. in the case of 1Password 8, it cannot even detect whether it is in a text input or not as everything is big ScrollViews. as far as kV is concerned there's no pingable text input fields in there. so as a final recourse, move the app to `911` where kV just enforces the key remapping as if everything was a text input. this is pretty bad stuff.

**Comment by godbout** [comment]
for the 1P launchbar, you "kinda" can make it work going up and down the list, although moves in the search box are still not working (not sure why):

https://user-images.githubusercontent.com/121373/163412254-c5bc4b98-83ab-41a5-9d15-2fcd8424e5aa.mp4

but for 1P itself, you guys seem to be doing some custom key logging thing? kV never receives the key presses as you can see below:

https://user-images.githubusercontent.com/121373/163412499-37668af4-4476-4f71-8910-52ca39ecbba4.mp4

it's like 1P grabs the keys, print them but never sends them back to macOS, so kV can never see them. not sure whether if it's you guys doing some custom handling or if it's the Secure Input Fields that are extra secured. never had the issue with 1P 7.

**Comment by mrjones2014** [comment]
I'm not sure but if I had to guess, that sounds like it could be intended behavior with secure input. I'm honestly not too worried about kindaVim working in 1Password 8.

99% of the time, I'm either using the browser extension which has its own nice shortcuts, or if I'm using the desktop app, I'm probably using the trackpad anyway. I wouldn't worry about that one.

**Comment by godbout** [comment]
so you got all the other ones working now?

i would recommend moving at least Slack and Signal into the `Key Mapping` one, for UX consistency. the reason it seems to be working when you put them in `Electron` is that kV checks if the inputs are of type TextField or TextArea (needed in order to get all the info to manipulate the text). in both Slack and Signal most of the text inputs are not, they're custom StaticText with no info available, so kV defaults to using `Key Mapping` instead. but some fields are indeed TextFields/TextAreas but with half the info missing. so you're gonna to have troubles with those ones. if you move those apps to `Key Mapping`, all the fields will behave the same. up to you.

**Comment by godbout** [comment]
ok great, thanks. now it's clear what i have to do. i'll still check the crash with Discord, and will tweak a few things. then back to a better sequence handling. thanks again.

**Comment by godbout** [comment]
> It also seems like sometimes entering normal mode in Discord sometimes crashes kindaVim entirely 😅

related to a fix implemented 16 days ago but that Xcode for whatever reason doesn't want to pull the latest master changes: https://github.com/godbout/Common/commit/f80ebe9bc55a6222df982f5256566dd8a774e3d0

the fix is to extra secure when Electron apps send back wrong data.

(writing for myself.)

**Comment by godbout** [comment]
was planning to release v17 today but kV is stuck on Apple's notarization servers. the delay is done and i'm honestly surprised how well it works. i'm almost ready to give up my `escape`!

hopefully out tomorrow. next will be #63.

**Comment by godbout** [comment]
kV17 out: https://github.com/godbout/kindaVim.theapp/releases/tag/17

**Comment by godbout** [comment]
@mrjones2014 let me know once you've found out it works well, so that we can close this issue. thank you!

**Comment by mrjones2014** [comment]
Works great, you can close the issue, thanks!
